File: framec_tests/csharp/EventHandler/EventHandler.cs
using FrameLang;
#nullable disable
namespace EventHandler
{

class EventHandler
{

    private EventHandlerCompartment _compartment_;
    private EventHandlerCompartment _nextCompartment_;



    public EventHandler(){


        // Create and intialize start state compartment.


        this._state_ = (int)EventHandlerState.S1;
        this._compartment_ = new EventHandlerCompartment(this._state_);
        this._nextCompartment_ = null;


        // Send system start event
        FrameEvent frameEvent = new FrameEvent(">", null);
        this._mux_(frameEvent);

    }

    // states enum
    private enum EventHandlerState
    {

        S1 = 0,
        S2 = 1
    }
    //====================== Multiplexer ====================//

    private void _mux_(FrameEvent e)
    {
        switch (this._compartment_.state)
        {
            case (int)EventHandlerState.S1:
                this._sS1_(e);
                break;
            case (int)EventHandlerState.S2:
                this._sS2_(e);
                break;
        }

        if( this._nextCompartment_ != null)
        {
            var nextCompartment = this._nextCompartment_;
            this._nextCompartment_ = null;
            if (nextCompartment._forwardEvent != null &&
               nextCompartment._forwardEvent._message == ">")
            {
                this._mux_(new FrameEvent( "<", this._compartment_.ExitArgs));
                this._compartment_ = nextCompartment;
                this._mux_(nextCompartment._forwardEvent);
            }
            else
            {
                this._doTransition_(nextCompartment);
                if (nextCompartment._forwardEvent != null)
                {
                    this._mux_(nextCompartment._forwardEvent);
                }
            }
            nextCompartment._forwardEvent = null;
        }
    }

    //===================== Interface Block ===================//

    public void LogIt(int x) {
        Dictionary<string,object> parameters = new Dictionary<string,object>();
        parameters["x"] = x;

        FrameEvent e = new FrameEvent("LogIt",parameters);
        this._mux_(e);
    }

    public void LogAdd(int a,int b) {
        Dictionary<string,object> parameters = new Dictionary<string,object>();
        parameters["a"] = a;

        parameters["b"] = b;

        FrameEvent e = new FrameEvent("LogAdd",parameters);
        this._mux_(e);
    }

    public int LogReturn(int a,int b) {
        Dictionary<string,object> parameters = new Dictionary<string,object>();
        parameters["a"] = a;

        parameters["b"] = b;

        FrameEvent e = new FrameEvent("LogReturn",parameters);
        this._mux_(e);
        return (int) e._return;
    }

    public void PassAdd(int a,int b) {
        Dictionary<string,object> parameters = new Dictionary<string,object>();
        parameters["a"] = a;

        parameters["b"] = b;

        FrameEvent e = new FrameEvent("PassAdd",parameters);
        this._mux_(e);
    }

    public int PassReturn(int a,int b) {
        Dictionary<string,object> parameters = new Dictionary<string,object>();
        parameters["a"] = a;

        parameters["b"] = b;

        FrameEvent e = new FrameEvent("PassReturn",parameters);
        this._mux_(e);
        return (int) e._return;
    }


    //===================== Machine Block ===================//

    EventHandlerCompartment compartment;


    private void _sS1_(FrameEvent e)
    {
        if (e._message == "LogIt")
        {
            this.log_do("x",(int)e._parameters["x"]);

            return;
        }
        else if (e._message == "LogAdd")
        {
            this.log_do("a",(int)e._parameters["a"]);
            this.log_do("b",(int)e._parameters["b"]);
            this.log_do("a+b",(int)e._parameters["a"] + (int)e._parameters["b"]);
            
            return;
        }
        else if (e._message == "LogReturn")
        {
            this.log_do("a",(int)e._parameters["a"]);
            this.log_do("b",(int)e._parameters["b"]);
            int r  = (int)e._parameters["a"] + (int)e._parameters["b"];
            this.log_do("r",r);
            e._return = r;

            return;

        }
        else if (e._message == "PassAdd")
        {


            compartment =  new EventHandlerCompartment((int)EventHandlerState.S2);

            compartment.StateArgs["p"] = (int)e._parameters["a"] + (int)e._parameters["b"];

            this._transition_(compartment);

            return;
        }
        else if (e._message == "PassReturn")
        {
            int r  = (int)e._parameters["a"] + (int)e._parameters["b"];
            this.log_do("r",r);


            compartment =  new EventHandlerCompartment((int)EventHandlerState.S2);

            compartment.StateArgs["p"] = r;

            this._transition_(compartment);
            e._return = r;

            return;

        }
    }

    private void _sS2_(FrameEvent e)
    {
        if (e._message == ">")
        {
            this.log_do("p",(int)this._compartment_.StateArgs["p"]);

            return;
        }
    }

    //===================== Actions Block ===================//

    public void log_do(string msg, int val)
    {

        string value = msg + "=" + val.ToString();
        this.tape.Add(value);
    }

    // Unimplemented Actions


    //===================== Domain Block ===================//

    public List<string> tape  = new List<string>();


    //=============== Machinery and Mechanisms ==============//

    private int _state_;

    private void _transition_(EventHandlerCompartment compartment)
    {
        this._nextCompartment_ = compartment;
    }

    private void _doTransition_(EventHandlerCompartment nextCompartment)
    {
        this._mux_(new FrameEvent("<", this._compartment_.ExitArgs));
        this._compartment_ = nextCompartment;
        this._mux_(new FrameEvent(">", this._compartment_.EnterArgs));
    }

    public string state_info(){
        return this._compartment_.state.ToString();
        }

}

//=============== Compartment ==============//

class EventHandlerCompartment
{

    public int state;

    public EventHandlerCompartment(int state)
    {
        this.state = state;
    }

    public Dictionary<string, object> StateArgs { get; set; } = new Dictionary<string, object>();
    public Dictionary<string, object> StateVars { get; set; } = new Dictionary<string, object>();
    public Dictionary<string, object> EnterArgs { get; set; } = new Dictionary<string, object>();
    public Dictionary<string, object> ExitArgs { get; set; } = new Dictionary<string, object>();
    public FrameEvent _forwardEvent = new FrameEvent();
}


// class EventHandlerController : EventHandler
// {
//         public EventHandlerController() : base()
//         {
//         }
// }
}

File: framec_tests/python/src/event_handler/event_handler.py
#Emitted from framec_v0.11.0



from framelang.framelang import FrameEvent



class FrameEvent:
    def __init__(self, message, parameters):
        self._message = message
        self._parameters = parameters
        self._return = None


class EventHandler:
    
    
    # ==================== System Factory =================== #
    
    def __init__(self):
        
         # Create and initialize start state compartment.
        
        next_compartment = None
        next_compartment = EventHandlerCompartment('__eventhandler_state_S1', next_compartment)
        self.__compartment = next_compartment
        self.__next_compartment = None
        
        # Initialize domain
        
        self.tape  = []
        
        # Send system start event
        frame_event = FrameEvent(">", None)
        self.__kernel(frame_event)
    
    # ==================== Interface Block ================== #
    
    def LogIt(self,x: int):
        parameters = {}
        parameters["x"] = x
        __e = FrameEvent("LogIt",parameters)
        self.__kernel(__e)
    
    def LogAdd(self,a: int,b: int):
        parameters = {}
        parameters["a"] = a
        parameters["b"] = b
        __e = FrameEvent("LogAdd",parameters)
        self.__kernel(__e)
    
    def LogReturn(self,a: int,b: int):
        parameters = {}
        parameters["a"] = a
        parameters["b"] = b
        __e = FrameEvent("LogReturn",parameters)
        self.__kernel(__e)
        return __e._return
    
    def PassAdd(self,a: int,b: int):
        parameters = {}
        parameters["a"] = a
        parameters["b"] = b
        __e = FrameEvent("PassAdd",parameters)
        self.__kernel(__e)
    
    def PassReturn(self,a: int,b: int):
        parameters = {}
        parameters["a"] = a
        parameters["b"] = b
        __e = FrameEvent("PassReturn",parameters)
        self.__kernel(__e)
        return __e._return
    
    # ===================== Machine Block =================== #
    
    # ----------------------------------------
    # $S1
    
    def __eventhandler_state_S1(self, __e, compartment):
        if __e._message == "LogIt":
            self.log_do("x",__e._parameters["x"])
            return
        elif __e._message == "LogAdd":
            self.log_do("a",__e._parameters["a"])
            self.log_do("b",__e._parameters["b"])
            self.log_do("a+b",__e._parameters["a"] + __e._parameters["b"])
            return
        elif __e._message == "LogReturn":
            self.log_do("a",__e._parameters["a"])
            self.log_do("b",__e._parameters["b"])
            r = __e._parameters["a"] + __e._parameters["b"]
            self.log_do("r",r)
            __e._return = r
            return
            
        elif __e._message == "PassAdd":
            next_compartment = None
            next_compartment = EventHandlerCompartment('__eventhandler_state_S2', next_compartment)
            next_compartment.state_args["p"] = __e._parameters["a"] + __e._parameters["b"]
            self.__transition(next_compartment)
            return
        elif __e._message == "PassReturn":
            r = __e._parameters["a"] + __e._parameters["b"]
            self.log_do("r",r)
            next_compartment = None
            next_compartment = EventHandlerCompartment('__eventhandler_state_S2', next_compartment)
            next_compartment.state_args["p"] = r
            self.__transition(next_compartment)
            __e._return = r
            return
            
    
    # ----------------------------------------
    # $S2
    
    def __eventhandler_state_S2(self, __e, compartment):
        if __e._message == ">":
            self.log_do("p",(compartment.state_args["p"]))
            return
    
    # ===================== Actions Block =================== #
    
    def log_do(self,msg: str,val: int):
        raise NotImplementedError
    
    # ==================== System Runtime =================== #
    
    def __kernel(self, __e):
        
        # send event to current state
        self.__router(__e)
        
        # loop until no transitions occur
        while self.__next_compartment != None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            
            # exit current state
            self.__router(FrameEvent( "<", self.__compartment.exit_args))
            # change state
            self.__compartment = next_compartment
            
            if next_compartment.forward_event is None:
                # send normal enter event
                self.__router(FrameEvent(">", self.__compartment.enter_args))
            else: # there is a forwarded event
                if next_compartment.forward_event._message == ">":
                    # forwarded event is enter event
                    self.__router(next_compartment.forward_event)
                else:
                    # forwarded event is not enter event
                    # send normal enter event
                    self.__router(FrameEvent(">", self.__compartment.enter_args))
                    # and now forward event to new, intialized state
                    self.__router(next_compartment.forward_event)
                next_compartment.forward_event = None
                
    
    def __router(self, __e):
        if self.__compartment.state == '__eventhandler_state_S1':
            self.__eventhandler_state_S1(__e, self.__compartment)
        elif self.__compartment.state == '__eventhandler_state_S2':
            self.__eventhandler_state_S2(__e, self.__compartment)
        
    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment
    
    def state_info(self):
        return self.__compartment.state
        

# ===================== Compartment =================== #

class EventHandlerCompartment:

    def __init__(self,state,parent_compartment):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment
    